p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/tools/P2RepositoryDataManipulator.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * Contributors:
+ *     Hannes Wellmann - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.tycho.p2maven.tools;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.maven.model.Repository;
+import org.eclipse.equinox.internal.p2.repository.helpers.RepositoryHelper;
+import org.eclipse.equinox.p2.core.ProvisionException;
+import org.eclipse.equinox.p2.internal.repository.tools.XZCompressor;
+import org.eclipse.equinox.p2.repository.IRepository;
+import org.eclipse.tycho.p2maven.repository.P2RepositoryKind;
+import org.eclipse.tycho.p2maven.repository.P2RepositoryManager;
+
+@Named
+@Singleton
+public class P2RepositoryDataManipulator {
+
+	public static ModifiedRepositoryDescriptor createDescriptor(Repository repositoryLocation,
+			P2RepositoryKind repositoryKind, Path outputLocation, boolean compress, boolean xzCompress,
+			boolean keepNonXzIndexFiles) throws IllegalArgumentException {
+
+		if (outputLocation != null) {
+			try {
+				URI location = RepositoryHelper.localRepoURIHelper(new URI(repositoryLocation.getUrl()));
+				if (location.getScheme().equals("file") && Files.isSameFile(outputLocation, Path.of(location))) {
+					outputLocation = null; // Signal in-place edit
+				} else if (!Files.isDirectory(outputLocation)) {
+					Files.createDirectories(outputLocation);
+				} else {
+					try (Stream<Path> files = Files.list(outputLocation);) {
+						if (files.findAny().isPresent()) {
+							throw new IllegalStateException("Output location is not empty: " + outputLocation);
+						}
+					}
+				}
+				repositoryLocation.setUrl(location.toString()); // reset (potentially) normalized URL
+			} catch (IOException e) {
+				throw new IllegalArgumentException("Not a writeable directory", e);
+			} catch (URISyntaxException e) {
+				throw new IllegalArgumentException("Not a valid location: " + repositoryLocation.getUrl(), e);
+			}
+		}
+		return new ModifiedRepositoryDescriptor(repositoryLocation, repositoryKind, outputLocation, compress,
+				xzCompress, keepNonXzIndexFiles);
+	}
+
+	public record ModifiedRepositoryDescriptor(Repository sourceRepository, P2RepositoryKind repositoryKind,
+			Path outputLocation, boolean compress, boolean xzCompress, boolean keepNonXzIndexFiles) {
+	}
+
+	@Inject
+	private P2RepositoryManager manager;
+
+	public void modifyRepositoryMetadata(ModifiedRepositoryDescriptor repository, String repositoryName,
+			List<String> propertiesToRemove, Map<String, String> propertiesToAdd)
+			throws ProvisionException, URISyntaxException, IOException {
```

Comment:
```suggestion
			throws ProvisionException, IOException {
```

I would catch `URISyntaxException` inside the method and convert it into an `IOException`. As an alternative don't pass in any data that requires paring of URIs but hand over proper URI objects in the first place.

Reply:
That's the biggest problem I see when using the maven Repository to model repositories as it models the location `URI` as String. This delays the surfacing of errors and requires to re-parse the String again and again (depending on the use-case). So ideally we would wrap the `Repository` immediately that only uses an URI or use another type (like `MavenRepositoryLocation`), but we discussed that already in
- https://github.com/eclipse-tycho/tycho/pull/5209#discussion_r2260505150

For now I'll keep using `Repository` for the user API and use `MavenRepositoryLocation` internally, this can then be unified in
- https://github.com/eclipse-tycho/tycho/pull/5222